### ISSUE TITLE
[Enhancement]: Use Action scheduler library for Woo Bulk actions

### ISSUE THREAD

**Issue by jonathan-dejong** [woocommerce/woocommerce#41239]
### Describe the solution you'd like

Since Action scheduler 3 came out a couple of years ago it's quite nice and stable with much less reliance on WP Cron. 
Bulk editing actions in WooCommerce have always been slow but the problem that incurs, combined with the fact that there's no limit to how many items one can list in the tables, is that we see a lot of users experiencing timeout issues. 
This is because they try to bulk edit too many products/orders/coupons at once and the update process for these isn't very fast, especially when throwing in a few Woo addon plugins that tack on to the save\_\* hooks.

I think it would make a lot of sense for WooCommerce to implement the action scheduler library for all bulk actions. 
This way there's no reasonable limit to how many products one could bulk edit at once without running into server limitation issues and it would also pave the way for making the Bulk edit feature a lot nicer.

Timeline-wise this makes sense to me given that WP core recently posted about the first iterations of taking the Post list tables into Gutenberg. Which I would assume also means we're no longer doing a GET request with a tonne of parameters in the URLs (another huge flaw in the original design if you ask me).

### Describe alternatives you've considered

there's a tonne of third party plugins trying to make bulk editing better but almost all of them have this same issue. They try to do things synchronously rather than asynchronously OR they absolutely spam the admin ajax with requests to get around it. 
For people who just need decent bulk editing that WC already provides, but without having to limit themselves to say max 100 products per iteration, this proposed solution would solve it for them.

### Additional context

Action scheduler library: 
[https://actionscheduler.org/](<https://actionscheduler.org/>)

**Comment by ObliviousHarmony** [comment]
I've taken a look at this issue and I have some ideas about how to achieve this.

The standard bulk-editing interface isn't filterable and [works by saving each post individually](https://github.com/WordPress/wordpress-develop/blob/fe2b33f27b8eef2b42771414a9464a3508cbca3a/src/wp-admin/includes/post.php#L613-L720). This is done using a `GET` request to `/wp-admin/edit.php`. As a consequence, there isn't really a way to make this flow asynchronous. WordPress' suggestion seems to be that [custom bulk edit behavior should be implemented using a `handle_bulk_actions-` filter](https://github.com/WordPress/wordpress-develop/blob/fe2b33f27b8eef2b42771414a9464a3508cbca3a/src/wp-admin/edit.php#L222-L223).

The `bulk_edit_posts()`[ WordPress function](https://github.com/WordPress/wordpress-develop/blob/fe2b33f27b8eef2b42771414a9464a3508cbca3a/src/wp-admin/includes/post.php#L504) lets you pass in the request data and that includes a list of posts to apply the edits to. We could save the request parameters and run the bulk update in batches. There aren't any points in `edit.php` that get filtered so I don't think there would be any reason we can't just directly call the function from an ActionScheduler job.

## Action Override

The fastest and probably most compatible approach would be to override the `action` parameter in the current form. With a custom action we can take advantage of the above filter to create an asynchronous job for the updates. When PHP receives multiple instances of the same non-array parameter it just overrides them. We can take advantage of that either through the use of a custom button "Background Update" or a hidden input that changes the action.

## Custom Editor

**To be clear**, the entire reason that the above filter exists is for developers to make *custom* editors. This is probably the recommended approach, however, it would mean jumping through hoops to try and maintain backward compatibility. There is always the possibility that JavaScript or other plugin filters inject content into the existing form with specific expectations. Assuming we didn't mind abandoning backward compatibility though, we could create a really awesome bulk editor this way.

It does feel a bit like the above solution with extra steps though.

## Conclusion

From my review, `bulk_edit_posts()` is *technically* able to be used asynchronously, however, it exists within the context of there being a logged in user. As far as I can tell, as long as we `switch_to_blog()` and `wp_set_current_user()`, I don't *think* that we would break anything by making the update asynchronous.

**Comment by jonathan-dejong** [comment]
Thank you for digging into this subject @ObliviousHarmony, I appreciate you!

The way you describe it is similar to how we've successfully done custom bulk actions before. 
Personally I don't think there should be a lot of reason to maintain backwards compatibility as synchronous saves in the original action. However I suppose it could be done by using a feature flag and if needed a separate action for this new approach and the flag simply adds/removes the actions as configured. 

However as mentioned in my original post, WP is bound to likely break all of that anyway when new table components or "views" are being adopted and if we already have this in place then it should be fairly easy to make the transition in this case. 

Anything I can do to help keep the momentum here?

**Comment by ObliviousHarmony** [comment]
> Personally I don't think there should be a lot of reason to maintain backwards compatibility as synchronous saves in the original action.

My concern with this approach is that the context for the execution changes in a way that could potentially break extensions that add functionality to bulk updates, or worse, elsewhere altogether. Right now, the bulk edits are done using the admin's active session and take place through the admin `edit.php` page. This does a full admin bootstrap and any extensions hooking into `bulk_update` may be relying on that context.

With that said, @prettyboymp had a really interesting suggestion. He proposed that we just spin up separate AJAX requests to `edit.php` for batches. I'm sure there are problems with that to consider too, but I wanted to share that idea too.

**Comment by prettyboymp** [comment]
> With that said, @prettyboymp had a really interesting suggestion. He proposed that we just spin up separate AJAX requests to `edit.php` for batches. I'm sure there are problems with that to consider too, but I wanted to share that idea too<br>

To expand on that, the idea was to catch the form submission for bulk updates and if there were more than X objects selected, to split the POST request into multiple smaller requests.

**Comment by jonathan-dejong** [comment]
I suppose that could work. That still leaves the user waiting for them all to complete, which could be very slow so you'd probably want to introduce some form of loader indicator. 
And it's susceptible to fail if the user navigates away from the page.
Basically it has the potential to make it less prone to hitting server limits but more prone to fail due to user behaviour.

In the original proposal above I actually mention this approach but as a thing others do that isn't actually asynchronous and stable.

> My concern with this approach is that the context for the execution changes in a way that could potentially break extensions that add functionality to bulk updates, or worse, elsewhere altogether. Right now, the bulk edits are done using the admin's active session and take place through the admin edit.php page. This does a full admin bootstrap and any extensions hooking into bulk_update may be relying on that context.

I will be bluntly honest here and say that this is a frustrating point of view I often encounter with these types of discussion in WC and personally I think it hurts the eco system more than helps it. 

Here's how it reads: Hypothetically someone's third party code could break if we improve this feature to take it to modern standards for processing large amounts of data and therefore we probably shouldn't do it or go for a subpar solution that introduces other end user frustrations instead. 

Is it not the job of developers to update their code around the software they build on? That's how almost every software with an eco system works, including WordPress itself.

**Comment by ObliviousHarmony** [comment]
> I will be bluntly honest here and say that this is a frustrating point of view I often encounter with these types of discussion in WC and personally I think it hurts the eco system more than helps it.

Honestly, it's an incredibly frustrating perspective to have to support. There are a number of "common sense" changes we could make to improve performance if we disregarded backward compatibility. I would love nothing more than to be able to radically modernize Woo. The problem is, when people update their store, they **need** it to continue working; downtime can often mean lost revenue,

WordPress makes it too easy to press the "Update" button and break your website. In some cases, changes can even result in fatals post-update that make it impossible to access the admin area. We need to be excessively careful with the changes that we make.

> Is it not the job of developers to update their code around the software they build on? That's how almost every software with an eco system works, including WordPress itself.

I could argue that some of that software also includes more explicit update processes aimed at power users. In my experience, a lot of the software that offers layperson-facing update options work to aggressively maintain backward compatibility. We can't restrict users from updating extensions (WP plugin guidelines) to new versions that don't explicitly declare compatibility with specific versions of WooCommerce.

In the interest of making sure I give a comprehensive reply here, we currently directly provide these edit saving hooks:

- `woocommerce_product_bulk_and_quick_edit`: Provided to let developers hook into WordPress's bulk editing and run custom code. We hook into this ourselves and a higher-priority hook can be used to make changes to state before we run our hooks.
- `woocommerce_product_quick_edit_save`: Provided to let developers hook into WooCommerce's quick edit saves at the end of the process.
- `woocommerce_product_bulk_edit_save`: Provided to let developers hook into WooCommerce's bulk edit saves at the end of the process.

As some specific examples of breaking compatibility with an ActionScheduler context, I am concerned with:

- No `current_user` will be set.
- `$_REQUEST`, `$_POST`, `$_GET`, `$_FILES`, and `$_COOKIE` superglobals will be empty.
- Any nonces that a hook expects will not exist.
- AS runs via an `admin-ajax.php` request and has the associated context. This is a particular problem for extensions and stores that optimize by changing the code they load and hooks they use when dealing with AJAX requests. The inverse is true as well, since any extension that relies on specific pages being visited (`edit.php`, in this case) will not work.
- AS `WP_CLI` commands are not ran with admin context by default, as I understand it. This means no admin files are loaded and `is_admin()` returns `false`.
- Any extension that uses global state _before_ WooCommerce's AS function has the opportunity to try and set it can  operate incorrectly.

In my opinion, using ActionScheduler for bulk edits would be a _significant_ break in backward compatibility. We could try storing and replicating the full request state, however, some of the big compatibility issues above can't be worked around. I don't really see a path forward using ActionScheduler for bulk editing.

While concurrent AJAX requests get us around the request state problems, they **don't** solve the AJAX context-related ones. I don't want to operate on an assumption here though, so I would want to do some research into extensions that use our bulk editing first. As for the UX issues, we can definitely find ways to solve that. This is something we could try and put behind a feature flag, however, I'm not sure how much of a priority it should be.

**Comment by pmcpinto** [comment]
Thanks for the report. I know this has been open for a long time, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

I'd like to get engineering to take a second look at this, so I've flagged it for developer feedback.

**Comment by gigitux** [comment]
Thanks for taking the time to share this suggestion. We appreciate the idea of using Action Scheduler for bulk actions. Given our current priorities and the backward compatibility concerns around changing how these actions run, we’re going to close this issue for now.
This doesn’t mean the suggestion isn’t valuable. We have limited time to work on improvements, and we may revisit it if our priorities change.
